### PR TITLE
Reverse proxy authentication to Woodpecker CI and Gitea

### PR DESCRIPTION
**Pull request by pboguslawski**
This mod adds option to authenticate user using HTTP header set by
reverse proxy. It forwards specified HTTP header with authenticated
username in requests to Gitea.

Requirements:

* Gitea must be configured for reverse proxy authentication and
  must accept HTTP header auth in API calls
  (https://github.com/go-gitea/gitea/pull/15119).

To enable set the following variables in woodpecker server
environment (woodpecker running behind reverse proxy):

* internal woodpecker server URL, i.e.:

  WOODPECKER_HOST_INTERNAL=http://192.168.1.100:8000/

* enable reverse proxy auth in woodpecker and forwarding auth
  header to gitea:

  WOODPECKER_GITEA_REV_PROXY_AUTH=true

* set name of header with authenticated username (set by
  reverse proxy), i.e.:

  WOODPECKER_GITEA_REV_PROXY_AUTH_HEADER=X-Forward-Username

Related: https://github.com/go-gitea/gitea/pull/15119
Author-Change-Id: IB#1107569

### CODE REVIEW

File: cmd/server/flags.go [hunk 1]
```diff
@@ -148,6 +156,16 @@ var flags = []cli.Flag{
 		Name:    "keepalive-min-time",
 		Usage:   "server-side enforcement policy on the minimum amount of time a client should wait before sending a keepalive ping.",
 	},
+	&cli.BoolFlag{
+		EnvVars: []string{"WOODPECKER_GITEA_REV_PROXY_AUTH"},
+		Name:    "gitea-rev-proxy-auth",
+		Usage:   "enable gitea authentication using HTTP header specified in WOODPECKER_GITEA_REV_PROXY_AUTH_HEADER",
+	},
+	&cli.StringFlag{
+		EnvVars: []string{"WOODPECKER_GITEA_REV_PROXY_AUTH_HEADER"},
```
[the remark this answers is not in the document]
Comment (pboguslawski) [reply cmd/server/flags.go:165]: Better to leave separate switch IHMO (easier for disable/enable bool temporarily for example). Simiar solution is already present in woodpecker (see WOODPECKER_GITHUB + WOODPECKER_GITHUB_URL for example) so this mod follows this idea.

**Comment by pboguslawski** [comment]
> I don't fully understand this PR, but why do you disable OAuth2 for the rev proxy auth? I'm not an expert with reverse proxies, but shouldn't it be possible to use OAuth2 with them?

If you use proxy for authentication, proxy is responsible for this and backend apps like gitea or woodpecker should trust it (user login passed in specified header). Because woodpecker must authenticate to gitea using HTTP call, it must also be able to forward header with user login to authenticate user in gitea. OAuth2 is different/separate auth logic and I don't see any point in trying to use it along with proxy auth header.

File: server/remote/gitea/gitea.go [hunk 2]
```diff
@@ -69,16 +74,63 @@ func New(opts Opts) (remote.Remote, error) {
...
+			return nil, err
+		}
+
+		// Since api does not return token secret, if drone token exists create new one.
+		resp, err := client.DeleteAccessToken("drone")
+		if err != nil && !(resp != nil && resp.StatusCode == 404) {
+			return nil, err
+		}
+		token, _, terr := client.CreateAccessToken(
+			gitea.CreateAccessTokenOption{Name: "drone"},
```
[the remark this answers is not in the document]
Comment (pboguslawski) [reply server/remote/gitea/gitea.go]: Fixed in 6315ebbbdc45184e52de6fe1790660d7ea76b755.

**Comment by pboguslawski** [comment]
> 1. Docs are missing, please add them.
> 
>     2. Please remove the chsnged copyright headers, we do this nowhere else.

Fixed.

**Comment by 6543** [comment]
well sorry that we dont come allong to review it ...

@pboguslawski what realy would help us is some docker-compose with such an auth-proxy in it ...
or an easy description how to set it up

so we can test it and play around with it

**Comment by pboguslawski** [comment]
Providing container infrastructure setup script/description for header auth is not easy, requires time and is outside the scope of this PR (feel free to create separate issue if required).

For simple tests consider setting `WOODPECKER_GITEA_REV_PROXY_AUTH_HEADER=X-Forward-Username` and sending request using curl like this
```
curl --header 'X-Forward-Username: mylogin' http://127.0.0.1:1234
```
or using your favourite proxy (i.e. apache, nginx, haproxy) in front of woodpecker + gitea backends with hardcoded `X-Forward-Username: mylogin` HTTP header injecting before passing request to backend (authenticated user simulation).

File: cmd/server/server.go [hunk 3]
```diff
@@ -100,6 +100,18 @@ func run(c *cli.Context) error {
...
+		log.Warn().Msg(
+			"WOODPECKER_HOST_INTERNAL should probably be publicly accessible (not localhost)",
+		)
+	}
+
+	if strings.HasSuffix(c.String("server-host-internal"), "/") {
+		log.Fatal().Msg(
+			"WOODPECKER_HOST_INTERNAL must not have trailing slash",
+		)
+	}
```
[the remark this answers is not in the document]
Comment (pboguslawski) [reply cmd/server/server.go:113]: This follows existing convention https://github.com/woodpecker-ci/woodpecker/blob/master/cmd/server/server.go#L100 so if change is required, should be applied separately.

**Comment by 6543** [comment]
CI failed :/ (lint issue)

**Pull request closed without merging**